gnoi/os/client.go

```diff
@@ -0,0 +1,81 @@
+/* Copyright 2020 Google Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package os
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/google/gnxi/gnoi/os/pb"
+	"google.golang.org/grpc"
+)
+
+// Client handles requesting OS RPCs.
+type Client struct {
+	client pb.OSClient
+}
+
+// ActivateErrorType represents the Type enum in ActivateError.
+type ActivateErrorType string
+
+// Enum representing possible ActivateErrorTypes.
+const (
+	ActivateUnspecified        ActivateErrorType = "UNSPECIFIED"
+	ActivateNonExistentVersion                   = "NON_EXISTENT_VERSION"
+)
+
+// ActivateError represents an error returned by the Activate RPC.
+type ActivateError struct {
+	error
+	ErrType ActivateErrorType
+	Detail  string
+}
+
+func (e *ActivateError) Error() string {
+	return fmt.Sprintf("%s: %s", e.ErrType, e.Detail)
+}
+
+// NewClient returns a new OS service client.
+func NewClient(c *grpc.ClientConn) *Client {
+	return &Client{client: pb.NewOSClient(c)}
+}
+
+// Activate invokes the Activate RPC for the OS service.
+func (c *Client) Activate(ctx context.Context, version string) error {
+	out, err := c.client.Activate(ctx, &pb.ActivateRequest{Version: version})
+	if err != nil {
+		return err
+	}
+	switch out.Response.(type) {
+	case *pb.ActivateResponse_ActivateOk:
+		return nil
+	case *pb.ActivateResponse_ActivateError:
+		res := out.GetActivateError()
+		errType := ActivateErrorType(res.GetType().String())
```

Comment:
these already exist in the proto generated code?
https://github.com/google/gnxi/blob/master/gnoi/os/pb/os.pb.go#L118

Reply:
I suppose I defined them in the os package so you wont have to import the proto generated code to do a type check in the main binary. As well as that these types contain strings so all you'd have to do is print it to the end user :) I can add another type for other error if you want

Reply:
oh I see what your plan is. I think you can actually keep the pb scoped to this file only. All the client here needs to do is to abstract the pb stuff from the main. The main only calls the actions and receives a (string) error. Why would it need more than that? =). This way you can simplify the error handling and avoid having to redefine new consts in this file every time the proto gets augmented.

```
	switch out.Response.(type) {
	case *pb.ActivateResponse_ActivateOk:
		return nil
	case *pb.ActivateResponse_ActivateError:
		res := out.GetActivateError()
		switch res.GetType() {
		case pb.ActivateError_UNSPECIFIED:
			return fmt.Errorf("Unspecified ActivateError: %s", res.GetDetail())
		case pb.ActivateError_NON_EXISTENT_VERSION:
			return fmt.Errorf("Non existent version: %s", version)
		default:
			return fmt.Errorf("Unknown ActivateError: %s", res.GetType())
		}
	}
```

Reply:
Thats true. Thanks!